features/atcryptoauth/mbedtls_atca_engine.cpp

```diff
@@ -0,0 +1,355 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* Mbed TLS includes */
+#include "mbedtls/asn1.h"
+#include "mbedtls/ecdsa.h"
+#include "mbedtls/pk_info.h"
+
+#include "ATCAFactory.h"
+#include "mbedtls_atca_engine.h"
+
+#define UNUSED(x) ((void)(x))
+
+/* Remove after rebase when this is available */
+#define MBEDTLS_ERR_PK_HW_ACCEL_FAILED (-1)
+
+/** Tell if can do the operation given by type.
+ *
+ *
+ * @param ctx       Key context.
+ * @param type      Target type
+ *
+ * @return          0 if context can't do the operations,
+ *                  1 otherwise.
+ */
+static int atca_can_do_func(const void *ctx, mbedtls_pk_type_t type)
+{
+    UNUSED(ctx);
+    return (MBEDTLS_PK_ECDSA == type);
+}
+
+/** Sign message.
+ *
+ *
+ * @param ctx       Key context
+ * @param md_alg    Algorithm that was used to hash the message
+ * @param hash      Message hash
+ * @param hash_len  Length of hash
+ * @param sig       Buffer that will hold the signature
+ * @param sig_len   Length of the signature written
+ * @param f_rng     RNG function
+ * @param p_rng     RNG parameter
+ *
+ * @retval          0 if successful, or appropriate Mbed TLS error code.
+ */
+static int atca_sign_func(void *ctx, mbedtls_md_type_t md_alg,
+                            const unsigned char *hash, size_t hash_len,
+                            unsigned char *sig, size_t *sig_len,
+                            int (*f_rng)(void *, unsigned char *, size_t),
+                            void *p_rng)
+{
+    ATCAKey * key = (ATCAKey *)ctx;
+    uint8_t rs[ATCA_ECC_SIG_LEN];
+    size_t rs_len;
+    int ret = 0;
+    ATCAError err = ATCA_SUCCESS;
```

Comment:
I'd rather initialize the error to indicate a bad state at the beginning. If the behaviour changes in the future and err will be written to only in a conditional code, the successful state will be propagated further.

Reply:
as in the (in)famous “goto fail” bug

Reply:
```err``` is not the return code here. It is defined on the top only to avoid compilation error due to ```goto``` statements. 
Though I will apply this comment to ```ret``` that is return code.